Enhance JavaScript code generation scopes with symbols

chengluyu opened the pull request:
This PR enhances scopes with symbols. The old code base lacks the clarification between lexical names (names used in MLscript) and runtime names (names used in generated JavaScript code), making translating declarations error-prune. Also, there is a need to memorize where symbols are declared and their kinds. For example, we want to know if `error` refers to the built-in `error` or a recent shadowed symbol.

Here is a brief change log.

- We added `ValueSymbol`, `TypeSymbol`, `BuiltinSymbol`, `StubValueSymbol`, `ClassSymbol`, `TraitSymbol` and `ParamSymbol` and related declaration methods in `Scope`.
- We refactored the result of `JSTestBackend`, making it more meaningful.
- Refering to a unimplemented symbol will not cause runtime errors. Instead, the test will output a short message. Also, the unimplemented status is cascadable. See dea229e.
- Provide an escape hatch to access built-in variables provided by Node.js. See a5ba631.

chengluyu committed:
WIP

chengluyu committed:
Merge branch 'mlscript' into scope-symbol

chengluyu committed:
Minor changes

chengluyu committed:
Finish new inheritance resolution but leave some bugs

chengluyu committed:
Make it works with basic examples

chengluyu committed:
All tests passed but there are some issues

- In `Polyfill.mls`, why `add` becomes `add1`. I think built-in symbols should not really occupy lexical scopes.
- In `BadAlias.mls`, why `Cannot allocate a runtime name starting with 'NonReg'`?
- In `Paper.mls`, `tryDiv` implementation allocates a new runtime name. This is wrong. The declaration of `ValueSymbol` should not override `StubValueSymbol`.
- In `RecursiveTypes.mls` if a symbol failed due to missing implementation, the symbol should also be marked as missing implementation.
- In `Ops.mls` why is the result 5? What was the generated code?
- In `SimpleMethods` class members should be able to see each other.
- In `Subsume.mls`, `ExprProp.mls` and `Group_2021_12_02.mls`, same problem as `Paper.mls`.
- In `BadTraits.mls` fix `class D0` . I should analyze trait base classes.
- In `MLTuples.mls` we still need to improve `withConstruct`.
- In `Methods.mls` why this?

    ```javascript
    ((H{}): D).(D.G)
    // becomes
    new H({})["D.G"];
    ```

- In `TypeDefs.mls` why crashed with `undeclared class Test111, did you call declareTypeDefs`?
- In `Traits.mls`, do we need to preserve additional fields? For example, `D { x = 0; y = "hello" }` where `D` has only field `x`.

chengluyu committed:
Add a missing test case file

chengluyu committed:
Merge branch 'mlscript' into scope-symbol

chengluyu committed:
Fix the infinite loop when substituting `TypeName(nothing)`

chengluyu committed:
Merge remote-tracking branch 'origin/mlscript' into scope-symbol

chengluyu committed:
Re-run tests after the merge commit

chengluyu committed:
Ignore cases of `Arr` and implement cases of `Subs`

chengluyu committed:
Generate runtime names with iterators

chengluyu committed:
Detect recursive unimplemented symbols and polish output

For example, in `MethodAndMatches.mls`:

```
def bar0: Base1[int] -> int -> Base1[int]
bar0 d1
bar0 d1 0
//│ bar0: Base1[int] -> int -> Base1[int]
//│     = <missing implementation>
//│ res: int -> Base1[int]
//│    = <no result>
//│      bar0 is not implemented
//│ res: Base1[int]
//│    = <no result>
//│      bar0 is not implemented
```

We show `<missing implementation>` for declarations and show
`<no result>` plus a reason for definitions which refer to
unimplmeneted symbols. The definition will become unimplemented.
So we can detect recursive unimplementation. Here is another
example from `RecursiveTypes.mls`.

```
rec def f x = (case x of {int -> f "ok" | string -> f (impossible x)}) + 1
//│ f: (int | string) -> int
//│  = <no result>
//│    impossible is not implemented

f 1
f "a"
//│ res: int
//│    = <no result>
//│      f and impossible are not implemented
//│ res: int
//│    = <no result>
//│      f and impossible are not implemented
```

Although `f` is defined, `f` refers to `impossible`, which is not implemented. So reference to `f` will be invalid too.

chengluyu committed:
Use `:escape` to allow access to stub values

chengluyu committed:
Add comments to `TemporaryVariableEmitter`

chengluyu committed:
Don't append numbers to identifers in some cases

- If the built-in symbol has not been accessed before, don't add
  numbers to the identifier. For example, if the built-in `add` is
  not used, declaring `add` will not has runtime name `add1`.
- [Bug] Define a declared stub value will use the same runtime
  name. I passed the wrong parameter and it is fixed now.

chengluyu committed:
Rename if the lexical name is a reserved word in JavaScript

chengluyu committed:
`this` should not be considered as a reserved word

chengluyu committed:
Move `inspectTerm` to a new file for helpers

chengluyu committed:
Refactor `JSTestBackend.generate` for better readability

chengluyu committed:
Remove the warning logger in `JSTestBackend`

chengluyu committed:
Handle arrays in `withConstruct` and add code assembly helpers

chengluyu committed:
Add test files missed in the previous commit

LPTK commented on shared/src/test/diff/mlscript/MLTuples.mls:
```diff
@@ -34,24 +44,24 @@ trait Hey: { x: int }
 //│ res: (1, 2, 3,) & hey | error
 //│    = [ 1, 2, 3 ]
 
 // TODO: fix `withConstruct` on arrays
```
Seems like this is now obsolete.
```suggestion
```

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -5,51 +5,45 @@ import scala.util.matching.Regex
+import mlscript.codegen.FreeSymbol
+import mlscript.codegen.ParamSymbol
+import mlscript.codegen.TypeSymbol
+import scala.collection.mutable.ArrayBuffer
```
```suggestion
import mlscript.codegen._
```

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -69,46 +63,46 @@ class JSBackend {
       val lamScope = Scope("Lam", patterns flatMap { _.bindings }, scope)
-      JSArrowFn(patterns, lamScope withTempVarDecls translateTerm(body)(lamScope))
+      JSArrowFn(patterns, lamScope.tempVars `with` translateTerm(body)(lamScope))
     // TODO: when scope symbols are ready, rewrite this
```
> TODO: when scope symbols are ready, rewrite this

So should this be rewritten now?

chengluyu replied:
It is a tiny optimization which turns built-in functions like `add 1 2` into `1 + 2`. I think it is not necessary now. So I just removed the TODO here.

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,208 @@
+// trait TypeSymbol {
+//   def baseType: Ls[Str] -> Type
+//   def isParametric: Bool
+// }
```
```suggestion
```

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,208 @@
+  /**
+    * A short name for the declaration used in debugging and error messages.
+    */
+  def shortName: Str
```
Seems unnecessary.

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,208 @@
+//   def isParametric: Bool
+// }
+
+abstract class LexicalSymbol extends RuntimeSymbol {
```
This and other base types should probably be `sealed`.

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,208 @@
+  override def shortName: Str = s"trait $lexicalName"
+}
+
+final case class ParamSymbol(override val lexicalName: Str, override val runtimeName: Str)
```
Seems unnecessary.

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,208 @@
+  def shortName: Str = s"free $lexicalName"
+}
+
+final case class TemporarySymbol(val runtimeName: Str) extends RuntimeSymbol
```
Seems unnecessary.

LPTK reviewed (changes requested):
Here's a summary of the change needs we identified during the meeting:

1. Fix eta expansion of classes shadowing a name

    ```scala
    Test = 1
    //│ Test: 1
    //│     = 1

    class Test
    //│ Defined class Test

    :js
    x = Test
    //│ // Query 1
    //│ globalThis.x = Test;
    //│ res = x;
    //│ // End of generated code
    //│ x: anything -> Test
    //│  = 1

    x 1
    //│ res: Test
    //│ Runtime error:
    //│   TypeError: x is not a function
    ```

2. Don't register type symbols as runtime names

3. What's the `location` variable for? Can it be removed?

I'm also going to commit and push the changes we made then.

LPTK committed:
Make code-gen simplifications and refactoring to make it more principled

chengluyu committed:
Fix some issues mentions in PR #81

- Rename `shortName` to `toString`.
- Turn `runtimeSymbols` to a `mutable.HashSet[Str]`, because its
  values are never used.
- Remove `TemporarySymbol`.
- Remove `ParamSymbol`.
- Improve inheritance hierarchy so that `TypeSymbol` and
  `TraitSymbol` won’t allocate runtime names. I made `ClassSymbol`
  and `TraitSymbol` inherit from `RuntimeSymbol`.

chengluyu committed:
Update shared/src/test/diff/mlscript/MLTuples.mls

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Commit a missing test file

chengluyu committed:
Organize imports in `JSBackend.scala`

chengluyu committed:
Remove an outdated TODO comment

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -1,18 +1,16 @@
-import mlscript.utils._, shorthands._
+import mlscript.codegen.Helpers._
+import mlscript.codegen._
+import mlscript.utils._
```
The `import mlscript.utils._, shorthands._` pattern is everywhere in the code base. No reason to make it different here. It sort if implicitly emphasizes that the two imports are related.

chengluyu replied:
Fixed in eb9dd5f.

chengluyu committed:
Rewrite stateful class symbol sorting

chengluyu committed:
Revert `import mlscript.utils._, shorthands._`

chengluyu committed:
Remove `LexicalSymbol.location`

chengluyu committed:
Keep recording if a stub symbol is shadowing another symbols

so that we can achieve definition -> declaration -> definition
cases in test files

chengluyu committed:
Implement `Ordered` for `ClassSymbol` to sort them stably

chengluyu committed:
Stop early if meet type aliases in `resolveClassBase`

chengluyu committed:
Simplify `resolveClassBase`

chengluyu committed:
Remove unused members in `ClassSymbol`

chengluyu committed:
Remove two more unused methods

chengluyu committed:
Merge remote-tracking branch 'origin/mlscript' into scope-symbol

chengluyu committed:
Add a missing test file

chengluyu committed:
Bring back error messages when use type names in expressions

chengluyu commented:
OK. I think I finished the cleanup.

LPTK commented on shared/src/main/scala/mlscript/codegen/Helpers.scala:
```diff
@@ -0,0 +1,67 @@
+    case Subs(arr, idx) => s"Subs(${inspect(arr)}, ${inspect(idx)})"
+  }
+
+  def topologicallySort[A: Ordering](relations: Ls[A -> A]): Iterable[A] = {
```
```suggestion
  def topologicalSort[A: Ordering](relations: Ls[A -> A]): Iterable[A] = {
```

LPTK replied:
Also, this general-purpose function should be moved to `shared/src/main/scala/mlscript/utils`, probably in a new file.

chengluyu replied:
Fixed in f499c7d.

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -243,287 +218,213 @@ class JSBackend {
+    val classSymbols = declareTypeDefs(typeDefs)
+    val (isolatedClasses, relations) = resolveInheritance(classSymbols)
+    val sortedClasses = try topologicallySort(relations) catch {
+      case e: RuntimeException => throw CodeGenError("cyclic inheritance involving")
```
The error message is incomplete! Just `"cyclic inheritance"` would actually be fine. We'll get a better error through the type anyway.

chengluyu replied:
Fixed in f499c7d.

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -243,287 +218,213 @@ class JSBackend {
+    val classSymbols = declareTypeDefs(typeDefs)
+    val (isolatedClasses, relations) = resolveInheritance(classSymbols)
+    val sortedClasses = try topologicallySort(relations) catch {
+      case e: RuntimeException => throw CodeGenError("cyclic inheritance involving")
```
> `case e: RuntimeException `

This is disgusting. It will swallow _any_ error and make it pass for a cyclic inheritance, even if it's another source or error. The better thing to do would be to throw a proper catchable exception in the topological sort function.

chengluyu replied:
Now we throw and catch `CyclicGraphError`.

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -243,287 +218,213 @@ class JSBackend {
+  /**
+    * Generate JavaScript code which targets MLscript test from the given block.
+    */
+  private def generate(pgrm: Pgrm)(implicit scope: Scope, allowEscape: Bool): JSTestBackend.TestCode = {
```
What does `allowEscape` do?! Document it?

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,168 @@
+  /**
+    * Fill up this field after resolving classes.
+    */
+  var baseClass: Opt[ClassSymbol] = N
```
Can you explain why `baseClass` and `body` need to be mutable variables? Why not have them as immutable parameters to the class?

chengluyu replied:
They were removed!

LPTK commented on shared/src/test/diff/codegen/TrickyShadowing.mls:
```diff
@@ -0,0 +1,240 @@
+//│ class add1 {}
+//│ // End of generated code
+
+// FIXME
```
```suggestion
```

LPTK commented on shared/src/test/diff/codegen/TrickyShadowing.mls:
```diff
@@ -0,0 +1,240 @@
+//│ x: anything -> Test
+//│  = [Function: x]
+
+// FIXME
```
```suggestion
```

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -204,8 +177,10 @@ class JSBackend {
+      JSMember(translateTerm(arr), translateTerm(idx))
+    case App(_, _) | _: Bind | _: Test =>
+      throw CodeGenError(s"cannot generate code for term ${inspect(term)}")
     case _ => die // FIXME "Exhaustivity analysis reached max recursion depth, not all missing cases are reported."
```
Let's try to fix this in this PR. Try with deeper max depth, and if it does not work maybe try to refactor a bit? Such as moving complicated patterns like `case App(App(App(Var("if"), tst), con), alt)` into the `App` case.

chengluyu replied:
I extract all cases of `App` into a new method. So exhaustivity check in `translateTerm` becomes easier.

LPTK replied:
Extracting into a different method seems unnecessary, but okay I guess.

LPTK reviewed (commented):
Otherwise LGTM! But before merging I'd like to make sure we really can't avoid that last bit of symbol mutability.

LPTK committed:
Add recursion test

chengluyu committed:
Fix minor issues mentioned in the PR

- Use specific error rather than `RuntimeException`
- Move topological sort to a new file
- Update cyclic inheritance error message

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -415,14 +415,19 @@ class JSTestBackend extends JSBackend {
+    *
+    * @param pgrm the program to translate
+    * @param scope the top-level scope
+    * @param allowEscape whether allow unimplemented values can be accessible
```
```suggestion
    * @param allowEscape whether to try executing code even if it refers to unimplemented definitions
```

LPTK committed:
WIP1

LPTK committed:
WIP2 lol

chengluyu committed:
Remove unused `fields` in `ClassSymbol`

chengluyu committed:
Remove stateful `baseClass` in `ClassSymbol`

chengluyu committed:
Remove stateful `body` in `ClassSymbol`

chengluyu committed:
Remove `translateClassDeclarations` and rewrite `declareTypeDefs`

chengluyu committed:
Make `resolveClassFields` static and extract `resolveClassBase`

chengluyu committed:
Reduce redundant calls of `resolveBaseClass`

chengluyu committed:
Merge scope-symbol with WIP1

chengluyu committed:
Merge "WIP2 lol" into scope-symbol

chengluyu committed:
Remove unused `TypeDefImpl`

chengluyu committed:
Clear Repro.mls

chengluyu committed:
Update shared/src/main/scala/mlscript/JSBackend.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Update shared/src/test/diff/codegen/TrickyShadowing.mls

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Update shared/src/test/diff/codegen/TrickyShadowing.mls

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Run tests and add changed files

LPTK commented on shared/src/main/scala/mlscript/utils/algorithms.scala:
```diff
@@ -17,8 +24,8 @@ object algorithms {
     }
-    val toPred = relations.foldLeft(SortedMap[A, Set[A]]()) { (acc, e) => 
-      acc + (e._1 -> acc.getOrElse(e._1, Set())) + (e._2 -> (acc.getOrElse(e._2, Set()) + e._1))
+    val toPred = edges.foldLeft(SortedMap.from(nodes.map { _ -> Set[A]() })) { (acc, e) => 
```
Faster:
```suggestion
    val toPred = edges.foldLeft(SortedMap.from(nodes.map { _ -> Set.empty[A] })) { (acc, e) => 
```

chengluyu committed:
Extract the translation of `App`

chengluyu commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -407,123 +244,157 @@ class JSBackend {
+  protected def sortClassSymbols(classSymbols: Ls[ClassSymbol]): Iterable[(ClassSymbol, Opt[ClassSymbol])] = {
+    // Cache base classes for class symbols.
+    // Note: it may include previously defined class symbols.
+    val baseClasses = Map.from(classSymbols.iterator.flatMap { derivedClass =>
+      resolveBaseClass(derivedClass.actualType).map(derivedClass -> _)
+    })
+    val sorted = try topologicalSort(baseClasses, classSymbols) catch {
+      case e: CyclicGraphError => throw CodeGenError("cyclic inheritance detected")
+    }
+    // We only need class symbols declared in current translation unit.
+    sorted.flatMap(sym => if (classSymbols.contains(sym)) S(sym -> baseClasses.get(sym)) else N)
```
Here we resolve the base class for each newly-defined classes and put them in a `Map`. The information is used immediately in `topologicalSort` (this is **the first consumer**). Note that `sortClassSymbols` returns all newly-defined classes with their base classes (`Iterable[ClassSymbol, Opt[ClassSymbol]`).

chengluyu commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -407,123 +244,157 @@ class JSBackend {
-          case Als => N
-        }
-      }
+    val classSymbols = declareTypeDefs(typeDefs)
+    val defStmts = sortClassSymbols(classSymbols).map { case (derived, base) =>
+      translateClassDeclaration(derived, base)(topLevelScope)
+    }.toList
```
After we got all sorted newly-defined classes (together with their base classes!), we pass them to `translateClassDeclaration`, which is **the second consumer** of base classes.

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -407,123 +244,157 @@ class JSBackend {
+    */
+  protected def sortClassSymbols(classSymbols: Ls[ClassSymbol]): Iterable[(ClassSymbol, Opt[ClassSymbol])] = {
+    // Cache base classes for class symbols.
+    // Note: it may include previously defined class symbols.
```
What is `it`? What is this comment trying to say?

chengluyu replied:
Fixed in 918dee7.

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,158 @@
+
+sealed trait TypeSymbol extends LexicalSymbol {
+  val params: Ls[Str]
+  val actualType: Type
```
These two params seem unused. Remove them? Or justify why we'll need them in the future.

chengluyu replied:
I noticed that @twitu is using two scopes (one for types and another for terms) in TypeScript code generation. So preserving `TypeSymbol` and related information could simplify the workflow there.

LPTK replied:
Ok, though he's not using symbols yet because they weren't really available. We'll look into using that info.

chengluyu replied:
OK. So I'm keeping it. We can remove them if it turns out to be useless.

LPTK commented on shared/src/main/scala/mlscript/codegen/Symbol.scala:
```diff
@@ -0,0 +1,158 @@
+    lexicalName: Str,
+    runtimeName: Str,
+    params: Ls[Str],
+    actualType: Type,
```
Why is the body type called the "actual" type? Very strange and nonstandard terminology. Please change.

chengluyu replied:
It was because there had been another member named `body`. Changed in `52324df`.

LPTK commented on shared/src/test/scala/mlscript/DiffTests.scala:
```diff
@@ -382,17 +434,22 @@ class DiffTests extends org.scalatest.funsuite.AnyFunSuite with org.scalatest.Pa
-                output(s"  $message")
+              case IllFormedCode(message) =>
+                totalRuntimeErrors += 1
+                output("Code generation met an error:")
```
`met` is a little strange here
```suggestion
                output("Code generation encountered an error:")
```

LPTK commented on shared/src/test/scala/mlscript/DiffTests.scala:
```diff
@@ -382,17 +434,22 @@ class DiffTests extends org.scalatest.funsuite.AnyFunSuite with org.scalatest.Pa
+              case Unimplemented(message) =>
+                output("Unable to execute the code:")
+                output(s"  ${message}")
+              case UnexpectedCrash(message) =>
```
This message should include the name of the error. For example try making it crash with a `MatchError` — it won't even say it's a match error.

chengluyu replied:
Fixed in 22ab793.

LPTK reviewed (approved):
Okay I think this is finally in a mergeable state! Apply my remaining comments and then feel free to squash-merge it.

chengluyu committed:
Use `Set.empty` instead of `Set` constructor

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Replace ”met“ with “encounter“ in an error message

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Run tests and commit changed files

chengluyu committed:
Rename `actualType` to `body` in `TypeSymbol`

chengluyu committed:
Show the error name when code generation crashes

chengluyu committed:
Clarify comments in `JSBackend.sortClassSymbols`

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -407,123 +244,157 @@ class JSBackend {
+      case e: CyclicGraphError => throw CodeGenError("cyclic inheritance detected")
+    }
+    // Their base classes might be class symbols defined in previous translation
+    // units. So we filter them here.
```
```suggestion
    // units. So we filter them out here.
```

chengluyu committed:
Fix a grammatical error in comments of `sortClassSymbols`

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>
